### PR TITLE
Various cleanups: Simplify product code

### PR DESCRIPTION
**Pull request by StephanTLavavej**
* `<exception>`: `throw_with_nested()` no longer needs `_Glued` to defend against the preprocessor.
  + After #3973, `_With_nested_v2<_Uty>` doesn't contain commas.
* Simplify: `ios_base::_Iostate{}` => `ios_base::goodbit`
  + It's just zero.
* Simplify `_Copy_unchecked()`'s `#ifdef __cpp_lib_concepts` logic.
  + Here's a case where a negative test allows us to considerably simplify the code by avoiding the need for two `#ifdef __cpp_lib_concepts` regions (which was hard to follow). Now the code reads much more naturally - in concepts mode, we need to handle different iterator/sentinel types.
* `<random>`: `discard_block_engine` can use its `result_type` typedef for `min` and `max`.
  + A constructor above is already using this typedef.
* Style: Avoid using an assignment expression as an index in `basic_string::_Eos()`.
  + We prefer to avoid expressions with lots of side effects, so we should update `_Mysize` before indexing. I'm using `_New_size` for the index since it's shorter.

### CODE REVIEW

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.
